### PR TITLE
Make database files' permissions configurable

### PR DESCRIPTION
**Pull request by xiaofeidu008**
We store some sensitive information in RocksDB. We want to give database files permission of 0600 rather than 0644 on Linux. So this diff adds an option to make it configurable.

### CODE REVIEW

**Comment by xiaofeidu008** [comment]
@ajkr There is a discussion thread in Rocksdb Users Group @workplace.
I copy the reason why `umask` doesn't solve our problem here: 

> Unfortunately, editing the umask also isn't an option for us as it will change the permissions of other non-RocksDB directories created for other functionality.

File: include/rocksdb/env.h [hunk 1]
```diff
@@ -80,6 +80,10 @@ struct EnvOptions {
   // If true, set the FD_CLOEXEC on open fd.
   bool set_fd_cloexec = true;
 
+  // If true, allow non owner read access for dbfile. Otherwise, non-owner
+  // has no access to the dbfile.
+  bool allow_dbfile_access_non_owner = true;
```
Comment (sagar0) [inline include/rocksdb/env.h]: `allow_non_owner_access` might be simpler and shorter?

File: util/env_posix.cc [hunk 2]
```diff
@@ -73,6 +73,14 @@ ThreadStatusUpdater* CreateThreadStatusUpdater() {
   return new ThreadStatusUpdater();
 }
 
+mode_t GetDBFileMode(bool allow_dbfile_access_non_owner) {
```
Comment (sagar0) [inline util/env_posix.cc]: Lets make this inline.

**Comment by sagar0** [comment]
Internal discussion: https://fb.facebook.com/groups/rocksdb/permalink/1789055621143045/

**Comment by ajkr** [comment]
Thanks for clarifying. Sorry I missed the internal discussion. I think this is a case where we can avoid introducing a rarely used DBOption. As proposed in the discussion, a setter on Env could change the file creation mask. It is fine to return NotSupported on non-PosixEnvs as we're not going to call that setter inside RocksDB, and other developers who want to use it can implement it for their Envs later.

**Comment by ajkr** [comment]
I was talking about adding a new function to `Env` class, which sets file mode. You can override it in PosixEnv to change the file creation mode from 0644 to 0600. Its default implementation can be returning NotSupported, i.e., you do not need to implement it for non-Linux platforms. The benefit is not adding rarely used options to DBOptions, which is an important goal.

**Comment by xiaofeidu008** [comment]
@ajkr @sagar0 I have created a new pull request which incorporates your feedback. See
https://github.com/facebook/rocksdb/pull/3709

**Comment by xiaofeidu008** [comment]
I will close this PR

**Pull request closed without merging**